cirq-core/cirq/transformers/routing/line_initial_mapper.py

```diff
@@ -0,0 +1,221 @@
+        """
+        circuit_graph: List[Deque['cirq.Qid']] = [deque([q]) for q in sorted(circuit.all_qubits())]
+        component_id: Dict['cirq.Qid', int] = {q[0]: i for i, q in enumerate(circuit_graph)}
+
+        def degree_lt_two(q: 'cirq.Qid'):
+            return any(circuit_graph[component_id[q]][i] == q for i in [-1, 0])
+
+        for op in circuit.all_operations():
+            if protocols.num_qubits(op) != 2:
+                continue
+
+            q0, q1 = op.qubits
+            c0, c1 = component_id[q0], component_id[q1]
+
+            # Keep track of partners for mapping isolated qubits later.
+            if q0 not in self.partners:
+                self.partners[q0] = q1
+            if q1 not in self.partners:
+                self.partners[q1] = q0
+
+            if not (degree_lt_two(q0) and degree_lt_two(q1) and c0 != c1):
+                continue
+
+            # Make sure c0/q0 are for the largest component.
+            if len(circuit_graph[c0]) < len(circuit_graph[c1]):
+                c0, c1, q0, q1 = c1, c0, q1, q0
+
+            # copy smaller component into larger one.
+            c1_order = (
+                reversed(circuit_graph[c1])
+                if circuit_graph[c1][-1] == q1
+                else iter(circuit_graph[c1])
+            )
+            for q in c1_order:
+                if circuit_graph[c0][0] == q0:
+                    circuit_graph[c0].appendleft(q)
+                else:
+                    circuit_graph[c0].append(q)
+                component_id[q] = c0
+
+        return sorted([circuit_graph[c] for c in set(component_id.values())], key=len, reverse=True)
+
+    def initial_mapping(self, circuit: 'cirq.AbstractCircuit') -> Dict['cirq.Qid', 'cirq.Qid']:
+        """Maps disjoint lines of logical qubits onto lines of physical qubits.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            a dictionary that maps logical qubits in the circuit (keys) to physical qubits on the
+            device (values).
+        """
+        qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = {}
+        circuit_graph = self._make_circuit_graph(circuit)
+        self.center = nx.center(self.device_graph)[0]
+
+        def next_physical(current_physical: 'cirq.Qid') -> 'cirq.Qid':
+            # Greedily map to highest degree neighbor that that is available
+            sorted_neighbors = sorted(
+                self.device_graph.neighbors(current_physical),
+                key=lambda x: self.device_graph.degree(x),
+                reverse=True,
+            )
+            for neighbor in sorted_neighbors:
+                if neighbor not in self.mapped_physicals:
+                    return neighbor
+            # If cannot map onto one long line of physical qubits, then break down into multiple
+            # small lines by finding nearest available qubit to the physical center
+            return self._closest_unmapped_qubit(self.center)
```

Comment:
Should we find a closest unmapped qubit to `current_physical` instead? If the outer loop of logical qubits is iterating over a connected line, then this would ensure that we find the closest node to the previous neighbour in line (i.e. `current_physical`), instead of forgetting about the line position and restarting from center?

Reply:
I'll keep note of this change and try it during benchmarking!